libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Steps/SwitchCondition.cs

```diff
@@ -74,30 +78,45 @@ protected override async Task<DialogTurnResult> OnRunCommandAsync(DialogContext
                 lock (this.Condition)
                 {
                     var engine = new ExpressionEngine();
-                    Expression condition = engine.Parse(this.Condition);
+                    conditionExpression = engine.Parse(this.Condition);
                     if (this.caseExpressions == null)
                     {
                         this.caseExpressions = new Dictionary<string, Expression>();
                         foreach (var c in this.Cases)
                         {
-                            var caseCondition = Expression.EqualsExpression(condition, engine.Parse(c.Value));
-                            // map of expression to steps
+                            // Values for cases are always coerced to string
+                            var caseCondition = Expression.ConstantExpression(c.Value);
+                            
+                            // Map of expression to steps
                             this.caseExpressions[c.Value] = caseCondition;
                         }
                     }
                 }
 
                 List<IDialog> stepsToRun = this.Default;
 
+                // Evaluate the condition expression, i.e. the left side of the switch equality
+                var (conditionEvaluation, conditionError) = this.conditionExpression.TryEvaluate(dc.State);
+
+                if (conditionError != null)
+                {
+                    throw new Exception($"Expression evaluation resulted in an error. Expression: {this.conditionExpression.ToString()}. Error: {conditionError}");
+                }
+
                 foreach (var caseCondition in this.Cases)
                 {
+                    var caseExpression = this.caseExpressions[caseCondition.Value];
 
-                    var (value, error) = this.caseExpressions[caseCondition.Value].TryEvaluate(dc.State);
-                    if (error != null)
+                    // Evaluate the constant expression for the right side of the equality
+                    var (value, error) = caseExpression.TryEvaluate(dc.State);
+
+                    if (conditionError != null)
                     {
-                        // Do what? 
+                        throw new Exception($"Expression evaluation resulted in an error. Expression: {caseExpression.ToString()}. Error: {error}");
                     }
-                    else if (((bool)value) == true)
+
+                    // Compare both expression results. The current switch case triggers if the comparison is true.
+                    if (conditionEvaluation.ToString().Equals(value.ToString(), StringComparison.OrdinalIgnoreCase))
```

Comment:
> if (conditionEvaluation.ToString().Equals(value.ToString(), StringComparison.OrdinalIgnoreCase)) [](start = 19, length = 97)

this doesn't seem right.  Why not Expression.EqualsExpression(condition, engine.Parse(c.Value)) ? ?